src/BSON/Iterator.c

```diff
@@ -263,10 +298,92 @@ static HashTable* php_phongo_iterator_get_properties(phongo_compat_object_handle
 	return php_phongo_iterator_get_properties_hash(object, false);
 }
 
+/* Iterator handlers */
+static void php_phongo_iterator_it_dtor(zend_object_iterator* iter)
+{
+	zval_ptr_dtor(&iter->data);
+}
+
+static int php_phongo_iterator_it_valid(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return intern->valid ? SUCCESS : FAILURE;
+}
+
+static zval* php_phongo_iterator_it_get_current_data(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	return php_phongo_iterator_get_current(intern);
+}
+
+static void php_phongo_iterator_it_get_current_key(zend_object_iterator* iter, zval* key)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	if (!php_phongo_iterator_key(intern, key)) {
+		// Exception already thrown
+		return;
+	}
+}
+
+static void php_phongo_iterator_it_move_forward(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	php_phongo_iterator_next(intern);
+}
+
+static void php_phongo_iterator_it_rewind(zend_object_iterator* iter)
+{
+	php_phongo_iterator_t* intern = Z_ITERATOR_OBJ_P(&iter->data);
+
+	php_phongo_iterator_rewind(intern);
+}
+
+static HashTable* php_phongo_iterator_it_get_gc(zend_object_iterator* iter, zval** table, int* n)
+{
+	*n     = 1;
+	*table = &iter->data;
+	return NULL;
+}
+
+static const zend_object_iterator_funcs php_phongo_iterator_it_funcs = {
+	php_phongo_iterator_it_dtor,
+	php_phongo_iterator_it_valid,
+	php_phongo_iterator_it_get_current_data,
+	php_phongo_iterator_it_get_current_key,
+	php_phongo_iterator_it_move_forward,
+	php_phongo_iterator_it_rewind,
+	NULL,
+	php_phongo_iterator_it_get_gc,
+};
+
+static zend_object_iterator* php_phongo_iterator_get_iterator(zend_class_entry* ce, zval* object, int by_ref)
+{
+	zend_object_iterator* iterator;
+
+	if (by_ref) {
+		phongo_throw_exception(PHONGO_ERROR_LOGIC, "An iterator cannot be used with foreach by reference");
+		return NULL;
+	}
```

Comment:
There is precedent for this in PHP itself, although they emit an `E_ERROR` in such cases. I opted to throw instead as it's more consistent how we expose errors.

We can't support iteration by reference as the BSON structures themselves are read-only, so assigning to the decoded value makes no sense.